### PR TITLE
Unified approach to metrics

### PR DESCRIPTION
**Pull request by Bez625**
I noticed we had two different approaches to metrics so tried to converge on the more recent approach. I have purposely deferred refactoring the client monitor to a later PR as it seemed to have a lot of impact and wanted to make the PRs more manageable.

### CODE REVIEW

**Comment by mcdee** [comment]
Please can you rebase this against current master and I'll take a look?  Thanks.

**Comment by Bez625** [comment]
> Please can you rebase this against current master and I'll take a look? Thanks.

Done! Thanks

File: services/attestationaggregator/standard/parameters.go [hunk 1]
```diff
@@ -101,6 +103,13 @@ func WithAggregateAndProofSigner(signer signer.AggregateAndProofSigner) Paramete
 	})
 }
 
+// WithChainTimeService sets the chain time service.
+func WithChainTimeService(service chaintime.Service) Parameter {
```
Comment (mcdee) [inline services/attestationaggregator/standard/parameters.go:107]: We should ensure that this is non-null in `parseAndCheckParameters`

File: services/attestationaggregator/standard/metrics.go [hunk 2]
```diff
@@ -119,20 +141,25 @@ func (s *Service) setupAttestationAggregationMetrics() error {
 	return nil
 }
 
-// AttestationAggregationCompleted is called when an attestation aggregationprocess has completed.
-func (s *Service) AttestationAggregationCompleted(started time.Time, slot phase0.Slot, result string) {
+func monitorAttestationAggregationCompleted(started time.Time, slot phase0.Slot, result string, chainTime chaintime.Service) {
```
Comment (mcdee) [inline services/attestationaggregator/standard/metrics.go:144]: I'm not a big fan of passing the service around in function calls.  It makes more sense to pass in either `startOfSlot` or `slotElapsed` as a variable here (and elsewhere this happens).

**Review by mcdee — CHANGES_REQUESTED** [review]
Apologies for the delay in reviewing this, a few items.

File: services/attestationaggregator/standard/service.go [hunk 3]
```diff
@@ -108,7 +115,9 @@ func (s *Service) Aggregate(ctx context.Context, data interface{}) {
 	duty, ok := data.(*attestationaggregator.Duty)
 	if !ok {
 		s.log.Error().Msg("Passed invalid data structure")
-		s.monitor.AttestationAggregationCompleted(started, 0, "failed")
+		// No duty so using 0 values for monitoring.
+		startOfSlot := s.chainTime.StartOfSlot(0)
+		monitorAttestationAggregationCompleted(started, 0, "failed", startOfSlot)
```
Comment (mcdee) [inline services/attestationaggregator/standard/service.go:120]: `startOfSlot` here can be replaced with a simple `0`.

File: services/attester/standard/attest.go [hunk 4]
```diff
@@ -39,29 +39,32 @@ func (s *Service) Attest(ctx context.Context, data interface{}) ([]*phase0.Attes
 
 	duty, ok := data.(*attester.Duty)
 	if !ok {
-		s.monitor.AttestationsCompleted(started, 0, len(duty.ValidatorIndices()), "failed")
+		// No duty so using 0 values for monitoring.
```
Comment (mcdee) [inline services/attester/standard/attest.go:42]: As per previous comment.

File: services/synccommitteemessenger/standard/service.go [hunk 5]
```diff
@@ -132,7 +136,9 @@ func (s *Service) Prepare(ctx context.Context, data interface{}) error {
 
 	duty, ok := data.(*synccommitteemessenger.Duty)
 	if !ok {
-		s.monitor.SyncCommitteeMessagesCompleted(started, 0, len(duty.ValidatorIndices()), "failed")
+		// No duty so using 0 values for monitoring.
+		startOfSlot := s.chainTimeService.StartOfSlot(0)
```
Comment (mcdee) [inline services/synccommitteemessenger/standard/service.go:140]: As per previous comment.

**Review by mcdee — CHANGES_REQUESTED** [review]
One minor tidy-up and I think we're good to go.

**Review by mcdee — APPROVED** [review]

**Pull request merged**